external/esp_idf_port/tcpip_adapter/tcpip_adapter_lwip.c

```diff
@@ -0,0 +1,1173 @@
+	struct netif *p_netif;
+
+	if (tcpip_if >= TCPIP_ADAPTER_IF_MAX || if_ip6 == NULL) {
+		return ESP_ERR_TCPIP_ADAPTER_INVALID_PARAMS;
+	}
+
+	p_netif = esp_netif[tcpip_if];
+	if (p_netif != NULL && netif_is_up(p_netif) && ip6_addr_ispreferred(netif_ip6_addr_state(p_netif, 0))) {
+		memcpy(if_ip6, &p_netif->ip6_addr[0], sizeof(ip6_addr_t));
+	} else {
+		return ESP_FAIL;
+	}
+	return ESP_OK;
+}
+
+esp_err_t tcpip_adapter_dhcps_option(tcpip_adapter_option_mode_t opt_op, tcpip_adapter_option_id_t opt_id, void *opt_val, uint32_t opt_len)
+{
+	void *opt_info = dhcps_option_info(opt_id, opt_len);
+
+	if (opt_info == NULL || opt_val == NULL) {
+		return ESP_ERR_TCPIP_ADAPTER_INVALID_PARAMS;
+	}
+
+	if (opt_op == TCPIP_ADAPTER_OP_GET) {
+		if (dhcps_status == TCPIP_ADAPTER_DHCP_STOPPED) {
+			return ESP_ERR_TCPIP_ADAPTER_DHCP_ALREADY_STOPPED;
+		}
+
+		switch (opt_id) {
+		case IP_ADDRESS_LEASE_TIME: {
+			*(uint32_t *) opt_val = *(uint32_t *) opt_info;
+			break;
+		}
+		case REQUESTED_IP_ADDRESS: {
+			memcpy(opt_val, opt_info, opt_len);
+			break;
+		}
+		case ROUTER_SOLICITATION_ADDRESS: {
+			if ((*(uint8_t *) opt_info) & OFFER_ROUTER) {
+				*(uint8_t *) opt_val = 1;
+			} else {
+				*(uint8_t *) opt_val = 0;
+			}
+			break;
+		}
+		case DOMAIN_NAME_SERVER: {
+			if ((*(uint8_t *) opt_info) & OFFER_DNS) {
+				*(uint8_t *) opt_val = 1;
+			} else {
+				*(uint8_t *) opt_val = 0;
+			}
+			break;
+		}
+		default:
+			break;
+		}
+	} else if (opt_op == TCPIP_ADAPTER_OP_SET) {
+		if (dhcps_status == TCPIP_ADAPTER_DHCP_STARTED) {
+			return ESP_ERR_TCPIP_ADAPTER_DHCP_ALREADY_STARTED;
+		}
+
+		switch (opt_id) {
+		case IP_ADDRESS_LEASE_TIME: {
+			if (*(uint32_t *) opt_val != 0) {
+				*(uint32_t *) opt_info = *(uint32_t *) opt_val;
+			} else {
+				*(uint32_t *) opt_info = DHCPS_LEASE_TIME_DEF;
+			}
+			break;
+		}
+		case REQUESTED_IP_ADDRESS: {
+			tcpip_adapter_ip_info_t info;
+			uint32_t softap_ip = 0;
+			uint32_t start_ip = 0;
+			uint32_t end_ip = 0;
+			dhcps_lease_t *poll = opt_val;
+
+			if (poll->enable) {
+				memset(&info, 0x00, sizeof(tcpip_adapter_ip_info_t));
+				tcpip_adapter_get_ip_info(ESP_IF_WIFI_AP, &info);
+				softap_ip = htonl(info.ip.addr);
+				start_ip = htonl(poll->start_ip.addr);
+				end_ip = htonl(poll->end_ip.addr);
+
+				/*config ip information can't contain local ip */
+				if ((start_ip <= softap_ip) && (softap_ip <= end_ip)) {
+					return ESP_ERR_TCPIP_ADAPTER_INVALID_PARAMS;
+				}
+
+				/*config ip information must be in the same segment as the local ip */
+				softap_ip >>= 8;
+				if ((start_ip >> 8 != softap_ip)
+					|| (end_ip >> 8 != softap_ip)) {
+					return ESP_ERR_TCPIP_ADAPTER_INVALID_PARAMS;
+				}
+
+				if (end_ip - start_ip > DHCPS_MAX_LEASE) {
+					return ESP_ERR_TCPIP_ADAPTER_INVALID_PARAMS;
+				}
+			}
+
+			memcpy(opt_info, opt_val, opt_len);
+			break;
+		}
+		case ROUTER_SOLICITATION_ADDRESS: {
+			if (*(uint8_t *) opt_val) {
+				*(uint8_t *) opt_info |= OFFER_ROUTER;
+			} else {
```

Comment:
remove the space, please check all the lines.

Reply:
done.